vllm/engine/async_llm_engine.py

```diff
@@ -733,43 +725,61 @@ async def encode(
         """
         async for output in self.process_request(
                 request_id,
-                prompt,
+                inputs,
                 pooling_params,
-                prompt_token_ids,
-                lora_request,
-                multi_modal_data,
+                lora_request=lora_request,
         ):
             yield output
 
-    async def process_request(
+    @overload
+    def process_request(
         self,
         request_id: str,
-        prompt: Optional[str],
+        inputs: PromptInputs,
+        params: SamplingParams,
+        lora_request: Optional[LoRARequest] = None,
+    ) -> AsyncIterator[RequestOutput]:
+        ...
+
+    @overload
+    def process_request(  # type: ignore[misc]
+        self,
+        request_id: str,
+        inputs: PromptInputs,
+        params: PoolingParams,
+        lora_request: Optional[LoRARequest] = None,
+    ) -> AsyncIterator[EmbeddingRequestOutput]:
+        ...
+
+    def process_request(
+        self,
+        request_id: str,
+        inputs: PromptInputs,
         params: Union[SamplingParams, PoolingParams],
-        prompt_token_ids: Optional[List[int]] = None,
         lora_request: Optional[LoRARequest] = None,
-        multi_modal_data: Optional[MultiModalData] = None,
     ) -> AsyncIterator[Union[RequestOutput, EmbeddingRequestOutput]]:
         """Common logic to process requests with SamplingParams or
         PoolingParams."""
-        arrival_time = time.time()
 
-        stream = await self.add_request(
-            request_id,
-            prompt,
-            params,
-            prompt_token_ids=prompt_token_ids,
-            arrival_time=arrival_time,
-            lora_request=lora_request,
-            multi_modal_data=multi_modal_data,
-        )
+        async def generator():
```

Comment:
I had to define these overloads to pass the type checker in my IDE (the mypy checker used in CI is currently useless for this because it ignores types that are defined across directories)

Reply:
> I see - why did we need to define the new generator function instead of leaving it as is? Ideally we shouldn't make changes unless they're necessary.

I wasn't able to overload async generators properly, so I had to wrap the generator inside the function. I'll try to apply the same pattern as `LLM._run_engine` to avoid the need for overloads.